dlc-manager/src/chain_monitor.rs

```diff
@@ -69,50 +68,156 @@ impl ChainMonitor {
     pub fn new(init_height: u64) -> Self {
         ChainMonitor {
             watched_tx: HashMap::new(),
+            watched_txo: HashMap::new(),
             last_height: init_height,
-            last_block_hashes: Vec::with_capacity(NB_SAVED_BLOCK_HASHES),
         }
     }
 
+    /// Returns true if the monitor doesn't contain any transaction to be watched.
+    pub fn is_empty(&self) -> bool {
+        self.watched_tx.is_empty()
+    }
+
     pub(crate) fn add_tx(&mut self, txid: Txid, channel_info: ChannelInfo) {
-        self.watched_tx.insert(txid, channel_info);
+        log::debug!("Watching transaction {txid}: {channel_info:?}");
+        self.watched_tx.insert(txid, WatchState::new(channel_info));
+
+        // When we watch a buffer transaction we also want to watch
+        // the buffer transaction _output_ so that we can detect when
+        // a CET spends it without having to watch every possible CET
+        if channel_info.tx_type == TxType::BufferTx {
+            let outpoint = OutPoint {
+                txid,
+                // We can safely assume that the buffer transaction
+                // only has one output
+                vout: 0,
+            };
+            self.add_txo(
+                outpoint,
+                ChannelInfo {
+                    channel_id: channel_info.channel_id,
+                    tx_type: TxType::Cet,
+                },
+            );
+        }
+    }
+
+    fn add_txo(&mut self, outpoint: OutPoint, channel_info: ChannelInfo) {
+        log::debug!("Watching transaction output {outpoint}: {channel_info:?}");
+        self.watched_txo
+            .insert(outpoint, WatchState::new(channel_info));
+    }
+
+    pub(crate) fn cleanup_channel(&mut self, channel_id: ChannelId) {
+        log::debug!("Cleaning up data related to channel {channel_id:?}");
+
+        self.watched_tx
+            .retain(|_, state| state.channel_id() != channel_id);
+
+        self.watched_txo
+            .retain(|_, state| state.channel_id() != channel_id);
     }
 
     pub(crate) fn remove_tx(&mut self, txid: &Txid) {
+        log::debug!("Stopped watching transaction {txid}");
         self.watched_tx.remove(txid);
     }
 
-    pub(crate) fn process_block(
-        &self,
-        block: &Block,
-        height: u64,
-    ) -> Vec<(Transaction, ChannelInfo)> {
-        let mut res = Vec::new();
-
+    /// Check if any watched transactions are part of the block, confirming them if so.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the new block's height is not exactly one more than the last processed height.
+    pub(crate) fn process_block(&mut self, block: &Block, height: u64) {
```

Comment:
FYI, in our fork we went a little bit further:

1. https://github.com/p2pderivatives/rust-dlc/commit/c3f88502fa41644b0bd576b34a2ddc9bc5fef745.
2. https://github.com/p2pderivatives/rust-dlc/commit/1010f95f930d98219f8c28c2854391fbbd081ff8.
3. https://github.com/p2pderivatives/rust-dlc/commit/e0588e879336c7f8e4708635fd0fd9bcc2eb7566.

But I'm not sure we can easily apply these patches because the `Blockchain` trait could no longer be implemented for `ElectrsBlockchainProvider` and `BitcoinCoreProvider`.

Reply:
Nevertheless, it might be worth looking at those changes after merging. I know you wanted to change how we monitor the blockchain anyway, so maybe only https://github.com/p2pderivatives/rust-dlc/commit/1010f95f930d98219f8c28c2854391fbbd081ff8 is relevant.

Reply:
Yeah I think better to take that later, but thanks for the pointers!